_posts/2019-08-05-cognito-azuread-sso.markdown

```diff
@@ -0,0 +1,91 @@
+
+![cognito in action](/img/cognito-flow.gif)
+
+A notable pain point is that when ever you need to amend the attributes associated with your user pool, then the whole user pool must be recreated, meaning a new user pool ID is generated. In turn, this means that the `Entity Id` within your Azure Enterprise Application - which contains the user pool ID - must be updated. If you are lucky, this is something you can do yourself, but that is unlikely to be the case in most large organisations.
+
+
+### Mapping attributes from Azure to Cognito attributes
+
+Typical attributes to configure in a user pool will be a principal name and email. Ensure that they are configured with `mutable` set to `true` so that Cognito can continue to synchronize the dynamic users it creates, with those users in Azure Active Directory. It's often a good idea to also include the `groups` attribute.
+
+It's an unfortunate trait of Cognito that it is not possible to amend custom attributes once created - including their `mutable` flag - so it's good to understand what attributes you need as early as possible, to avoid having to re-create the Cognito User Pool over and over.
+
+When using Cognito as a facilitator for single sign on with Azure Active Directory, you will probably want to use the principal's directory based security groups. Note that it is also possible to manage your own user groups in Cognito, though many organisations prefer to manage their authentication and authorisation with a single identity provider.
+
+Adding custom attributes is as simple as defining them in your Cognito User Pool. From there, you can add mappings to them from your identity provider's available claim types - these are available in the metadata xml file provided by Azure AD when your Enterpise Application is created. Remember - make your cognito attributes mutable if you want to map them from your identity provider.
+
+## Bringing it all together
+
+There can be various stumbling blocks along the way when bringing together a single sign on solution with Cognito and Azure AD. Here are some things to watch out for:
+
+* Keep the SAML Identifier (Entity Id) correct and up to date in your Azure Enterprise Application
+* Ensure your SAML Reply URL is appropriate for your Cognito domain (e.g. https://**YOUR-COGNITO-DOMAIN**.auth.**AWS-REGION**.amazoncognito.com/saml2/idpresponse ) in your Azure Enterprise Application
+* If you are not getting AD groups in your token, ensure you have enabled them in the _User Attributes & Claims_ screen of your Azure Enterprise Application
+* If you get Cognito errors about redirect_uri, ensure your Cognito App Client settings _Reply URLs_ are correct (e.g. https://**your-load-balancer-address-or-domain**/oauth2/idpresponse) - note that you can include multiple reply and logout URLs by comma separating the URLs
+
+Once you've made this work one time, you'll find it easy to share among your team and clients in the future, and repeat over and again. 
```

Comment:
clients? Should that be customers or 3rd parties?

Reply:
ok let's go with customers